### PR TITLE
expose profiles by adding pprof to controllers

### PR DESCRIPTION
**Pull request by QuantumEnigmaa**
Towards https://github.com/giantswarm/giantswarm/issues/29151

### CODE REVIEW

File: helm/logging-operator/values.schema.json [hunk 1]
```diff
@@ -40,6 +40,9 @@
         "managementCluster": {
             "type": "object",
             "properties": {
+                "insecureCA": {
+                    "type": "boolean"
+                },
```
Comment (marieroque) [inline helm/logging-operator/values.schema.json:45]: Why that change ?

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+
+- Expose profiles in the controller.
+- Add conditional profiling annotations in the deployment.
```
Comment (marieroque) [inline CHANGELOG.md:13]: I'd merge that 2 points in one

File: main.go [hunk 3]
```diff
@@ -20,6 +20,10 @@ import (
 	"flag"
 	"os"
 
+	"net/http"
+	"net/http/pprof"
+	_ "net/http/pprof" //nolint: all
```
Comment (marieroque) [inline main.go:25]: Why those 3 imports, the last one is not enough ?

File: helm/logging-operator/values.schema.json (hunk 1 above)
Comment (QuantumEnigmaa) [reply helm/logging-operator/values.schema.json:45]: It's not a change I introduced myself, it just appears that some previous changes to the values file were not forwarded to the values schema

File: main.go (hunk 3 above)
Comment (QuantumEnigmaa) [reply main.go:25]: Oh yeah you're right my bad

File: CHANGELOG.md (hunk 2 above)
Comment (QuantumEnigmaa) [reply CHANGELOG.md]: Done

**Comment by hervenicol** [comment]
I did not test, so I'm guessing things here. Can you confirm I understood it well?

This PR adds support for profiling to logging-operator.

## In the binary:
It does so by adding a `/debug/pprof/` endpoint to the existing http server (that was only used to serve metrics until this PR).
The `/metrics` endpoint is still available, did not change.
This is the standard way of exposing profiles in golang, and this has no performance impact when nobody queries this endpoint. So there's no need for config parameters to enable/disable this.

## In the helm chart

We have to add some annotations to the pod so that pyroscope discovers and consumes the profiling data from logging-operator.
The annotations will be added with `.Values.profiling.enabled`.
When doing this, if pyroscope is running on the cluster, it will collect continuous profiling data from logging-operator.

**Comment by QuantumEnigmaa** [comment]
You sumed it perfectly ! The only thing I'm not entirely sure if is whether exposing the profiling endpoint implies an additional resources cost. I guess it shouldn't though

File: main.go [hunk 4]
```diff
@@ -86,12 +89,22 @@ func main() {
 	opts.BindFlags(flag.CommandLine)
 	flag.Parse()
 
+	mux := http.NewServeMux()
```
Comment (QuentinBisson) [inline main.go:92]: Have you considered using this https://github.com/kubernetes-sigs/controller-runtime/pull/1943/files ?

**Review by QuentinBisson — APPROVED** [review]

**Comment by hervenicol** [comment]
> You sumed it perfectly ! The only thing I'm not entirely sure if is whether exposing the profiling endpoint implies an additional resources cost. I guess it shouldn't though

They say that's "often with minimal overhead" (https://grafana.com/docs/pyroscope/latest/introduction/profiling/#continuous-profiling-1).
I guess in such a simple app there's not much concern, but could be a problem on more complex and resource-intensive apps.

File: main.go [hunk 5]
```diff
@@ -86,16 +86,27 @@ func main() {
...
+	mux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
+	mux.HandleFunc("/debug/pprof/trace", pprof.Trace) */
+
 	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
 
 	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
 		Scheme: scheme,
 		Metrics: server.Options{
 			BindAddress: metricsAddr,
+			/* ExtraHandlers: map[string]http.Handler{
```
[the remark this answers is not in the document]
Comment (QuantumEnigmaa) [reply main.go:104]: It was just during the testing phase, I'll remove those now :)

**Comment by QuantumEnigmaa** [comment]
Tested and working on `gaia` :)

File: main.go [hunk 6]
```diff
@@ -96,6 +98,7 @@ func main() {
 		HealthProbeBindAddress: probeAddr,
 		LeaderElection:         enableLeaderElection,
 		LeaderElectionID:       "5c8bbafe.x-k8s.io",
+		PprofBindAddress:       profilesAddr,
```
Comment (hervenicol) [inline main.go:101]: Oh, it's managed directly by `controller-runtime`! :exploding_head:

**Review by hervenicol — APPROVED** [review]
LGTM

**Pull request merged**